dev/Kozani/KozaniRemoteManager/KozaniDvcServer.cpp

```diff
@@ -0,0 +1,734 @@
+                }
+            
+                break;
+
+            case Dvc::ProtocolDataUnit::AppTerminationNotice:
+                RETURN_IF_FAILED(ProcessAppTerminationNotice(pdu));
+                break;
+
+            default:
+                // ToDo: Unsupported PDU type. Return GenericResult message with failure code to client side.
+                break;
+        }
+
+        return S_OK;
+    }
+    CATCH_RETURN()
+
+    void KozaniDvcServer::ProcessActivateAppRequestThreadFunc(_In_ Dvc::ProtocolDataUnit pdu) noexcept try
+    {
+        bool appActivatedAndTracked{};
+        std::string errorMessage;
+        HRESULT hr = LOG_IF_FAILED(ProcessActivateAppRequest(pdu, appActivatedAndTracked, errorMessage));
+
+        // If ProcessActivateAppRequest failed but appActivatedAndTracked is true, it means there was failure after the app was activated and 
+        // its process lifetime tracked. Such failure is benign and we should not send failure ActivateAppResult to client.
+        if (FAILED(hr) && !appActivatedAndTracked)
+        {
+            SendActivateAppResult(pdu.activity_id(), hr, 0, false, errorMessage);
+
+            // ERROR_ALREADY_EXISTS is due to the activity Id being the same as an existing activity already tracked in the m_activityMap. 
+            // Do not remove it from the map. 
+            if (hr != HRESULT_FROM_WIN32(ERROR_ALREADY_EXISTS))
+            {
+                auto lock{ g_connectionManager.m_activityMapLock.lock_exclusive() };
+                if (g_connectionManager.m_closing)
+                {
+                    return;
+                }
+
+                g_connectionManager.m_activityMap.erase(pdu.activity_id());
+            }
+        }
+    }
+    CATCH_LOG_RETURN()
+
+    HRESULT KozaniDvcServer::ProcessActivateAppRequest(
+        _In_ Dvc::ProtocolDataUnit& pdu, 
+        _Out_ bool& appActivatedAndTracked, 
+        _Out_ std::string& errorMessage) noexcept try
+    {
+        appActivatedAndTracked = false;
+        errorMessage.clear();
+
+        Dvc::ActivateAppRequest activateAppRequest;
+        if (!activateAppRequest.ParseFromString(pdu.data()))
+        {
+            errorMessage = "Failed to parse ActivateAppRequest message";
+            RETURN_WIN32(ERROR_INVALID_DATA);
+        }
+
+        if (!IsActivationKindSupported(activateAppRequest.activation_kind()))
+        {
+            errorMessage = "Unsupported activation kind: " + std::to_string(activateAppRequest.activation_kind());
+            RETURN_WIN32(ERROR_NOT_SUPPORTED);
+        }
+
+        {
+            // Acquire lock before adding the activity Id to the m_activityMap but do not hold the lock while activating the app as 
+            // that can take some time and we do not want one app activation to block others.
+            auto lock{ g_connectionManager.m_activityMapLock.lock_exclusive() };
+            RETURN_HR_IF(HRESULT_FROM_WIN32(ERROR_INVALID_STATE), g_connectionManager.m_closing);
```

Comment:
This can happen when the DVC stack has unrecoverable failure, like failing to create a DVC channel handle. With these kinds of failure, the COM server will decrement the object ref count so that the server will be shutting down, and destructor of the g_connectionManager will be called, and m_closing set to true. Unlikely to hit this code when that happens but better guard against that logically. ERROR_INVALID_STATE is the right error code here.